README.md
# NgbrxPaginator

You have an application that uses ngrx to store your data.
NgbrxPaginator gives you a really easy way to paginate your data with filters.

This package is in **BETA** and is developped with angular@16, ngrx@16 and ng-bootstrap@15.

The code is rather simple and if you want to join, you are more than welcome.

**This package does not deal with backend pagination.**




### Demo https://collorg.github.io/ngbrx-paginator-demo

### Installation

```bash
npm i ngbrx-paginator
```

ngbrx-paginator depends on [ngrx](https://ngrx.io/) and [ng-bootstrap](https://ng-bootstrap.github.io).

## Usage

The following example is extracted from the [departement](./projects/test-paginator/src/app/departement) module of the test application.

Suppose you have a `departement` [sic] module that has a slice of your store containing all the departments of France. The module provides a `departements` component that displays an unpaginated list of the departments:

departements.component.ts
```ts
import { Component } from '@angular/core';
import { Observable } from 'rxjs';
import { Departement } from '../departement.model';
import { Store } from '@ngrx/store';
import * as fromStore from '../departement.reducer';

@Component({
  selector: 'app-departements',
  templateUrl: './departements.component.html',
  styleUrls: ['./departements.component.css']
})
export class DepartementsComponent {
  collection$: Observable<Departement[]> = this.store.select(fromStore.selectAll);

  constructor(
    private store: Store<fromStore.State>
  ) { }

}
```

departements.component.html
```html
<div class="card">
  <div class="card-header sticky-top">
    <h3 clas="row">
      <span class="col-md">Liste des départements</span>
    </h3>
  </div>
  <div class="card-body">
    <div class="list-group">
      <div class="list-group-item" *ngFor="let item of collection$ | async">
        {{ item.code }} {{ item.nom }}
      </div>
    </div>
  </div>
</div>
```

![unpaginated departments](images/unpaginated-departements.png)

First, add the `NgbrxPaginatorModule` to your [departement.module.ts](./projects/test-paginator/src/app/departement/departement.module.ts) dependencies:

```ts
import * as fromDepartement from './departement.reducer';
import { NgbrxPaginatorModule } from 'ngbrx-paginator';

@NgModule({
  [...]
  imports: [
    [...]
    NgbrxPaginatorModule.forFeature({
      'Departement/Pagination': // The name of the paginator must be unique for the application.
      {
        // Optional
        filters: { // You can provide more than one filter by paginator.
          'Nom': { filter: fromDepartement.byName },
          'Code': {
            filter: fromDepartement.byCode,
            inactivate: true // Optional: not displayed when the page is loaded for the first time.
          },
          'Régions/COM': {
            filter: fromDepartement.byRegion,
            values: fromDepartement.selectRegions // Optional: provides an observable of the select values object.
          }
        },
        pageSizeOptions: [10, 20, 30] // Defaults to [5, 10, 25, 100]
      } 
    }),
    [...]
  ],
```

The object provided to the forFeature static method describes the paginators for this module.
Each paginator is an object identified by a unique name in the application with the following properties:
    
* `filters?`: (optional) an object of type `{ [key: string]: { filter, values?, inactivate? } }`.

  Each filter is identified by a key which will be used as a placeholder for the input or the default value for the select. To each key is associated an object with the following properties:

  * `filter`: a filter function that takes two parameters `(items: Data[], query: string)` and returns `Data[]`;
  * `values`: (optional) a MemoizedSelector. If provided, a HTML select element is displayed with the values instead of the input. The object's type is: { [key: string]: string;[key: number]: string }.
  * `inactivate`: (optional) boolean. If set to true the filter is inactivated when the paginator is loaded for the first time.
* `pageSizeOptions?`: (optional) defaults to [5, 10, 25, 100].

You have done the most important part of the job. Now, in your component class, add the attribute `key` to specify the name of the paginator, and use `NgbrxPaginationService` to paginate/filter your collection:

```ts
import { Component } from '@angular/core';
import { Observable } from 'rxjs';
import { Store } from '@ngrx/store';
import * as fromStore from '../departement.reducer';

import { Departement } from '../departement.model';
import { NgbrxPaginatorService } from 'ngbrx-paginator';

@Component({
  selector: 'app-departements',
  templateUrl: './departements.component.html',
  styleUrls: ['./departements.component.css']
})
export class DepartementsComponent {
  key = 'Departement/Pagination'; // same as in NgbrxPaginatorModules.forFeature
  collection$: Observable<Departement[]> = this.fromDepartement.selectAll;
  paginatedCollection$: Observable<Departement[]> = this.paginationService.setPaginator(this.key);

  constructor(
    private store: Store<fromStore.State>,
    // inject NgbrxPaginatorService
    private paginationService: NgbrxPaginatorService
  ) { }

}
```

Finally, add the `ngbrx-paginator` component and the optional `ngbrx-paginator-filter-desc` component in your [template](./projects/test-paginator/src/app/departement/departements/departements.component.html):

```html
<div class="card">
  <div class="card-header sticky-top">
    <h3>
      Liste des départements
    <!-- add (optional) --> <ngbrx-paginator-filter-desc class="right" [key]="key"></ngbrx-paginator-filter-desc>
    </h3>
    <!-- add --> <ngbrx-paginator [key]="key"></ngbrx-paginator>
  </div>
  <div class="card-body">
    <div class="list-group">
      <div class="list-group-item" *ngFor="let item of paginatedCollection$ | async"> <!-- replace collection$ -->
        {{ item.code }} {{ item.nom }}
      </div>
    </div>
  </div>
</div>
```

Voilà ! You now have a paginated list of `departements` with filters:

![départements](images/departements.png)

CHANGELOG
# release 0.3.7 (2023-10-19)

* Bump @angular/cli from 16.2.0 to 1.2.7, @angular/core from 16.2.0 to 16.2.10 and @ngrx/store from 16.2.0 to 16.3.0 (a16d589)
* Bump @babel/traverse from 7.22.17 to 7.23.2 (#5) (6689de3)

# release 0.3.6 (2023-10-04)

* Fix #4. Clear filters values on clearFilters action. (5c984f4)

# release 0.3.5 (2023-09-29)

* Add delay to trigger filter query. (a23f2dc)

# release 0.3.4 (2023-09-27)

* Add page-selector component. Remove use of subscribe. (caeab21)
* Scroll to top of the window on  page change. (50c213b)
* [UI] Improve placement of the filter activation button. (ae3a4e2)
* Do not alter the original data$ observable. (6c0886d)

# release 0.3.3 (2023-09-26)

* Add button clear filters with filter-desc component. (29795d9)

# release 0.3.2 (2023-09-25)

* Use ChangeDetectionStrategy.OnPush for ngbrx-paginator.component. (01a151d)

# release 0.3.1 (2023-09-21)

* [ngbrx-paginator-component] Fix order in select values (remove keyvalue pipe). (e481912)

# release 0.3.0 (2023-09-20)

* Replace suffix with extension. (13200e5)
* [BREAKING CHANGE] dataSelector removed from NgbrxPaginatorModule.forFeature paginator argument. Allow reuse of a paginator. (938ddc3)
* [BREAKING CHANGE] Rename allDataSelector by dataSelector. (3058e57)
* Move initialization of the paginator into ngbrx-paginator component. (936543d)
* [breaking change] The values of a filter must be of type: { [key: string]: string;[key: number]: string }. (39ac042)
* refactor filter description component. (b110675)

## Breaking changes

### 

* The object provided to the forFeature static method has changed:
  * it now takes only two optional attributes: `filters` and `pageSizeOptions`: `allDataSelector` has been removed;
  * the optional `values` are now of type: `{ [key: string]: string;[key: number]: string }`. It was previously a list.

* The paginator is initialized in the component with `NgbrxPaginatorService.setPaginator<M>(key: string, data$: Observable<M[]>, extension?: string): Observable<M[]>`.
  This method returns the paginated data observable (see the example in the README.)
* `NgbrxPaginatorComponent` now takes an optional input `extension` to allow multiple usage of the same paginator in different contexts.
  `NgbrxPaginatorDescComponent` has also this new optional input.

# release 0.2.7 (2023-08-23)

* Add inactivate option to filter. (48db418)

# release 0.2.6 (2023-08-23)

* Remove selectedFilters (replaced by activatedFilters). (4434937)
* Move filter selector. (78aa1e0)

# release 0.2.5 (2023-08-22)

* Fix bug with old localStorage. (d5562a0)

# release 0.2.4 (2023-08-22)

* Add a filter selector to activate/inactivate a filter. (e5ed558)
* Refresh yarn install. (6e43013)
* Bump @angular and @ngrx to 16.2.0 (a246837)

# release 0.2.3 (2023-08-03)

* yarn install --upgrade (6f66a16)
* [Fix #3] Don't use localStorage if something has changed in NgbrxPaginatorModule.forFeature({key: ...}) (3b893ff)

# release 0.2.2 (2023-07-31)

* Fix bug when localStorage was not present. (82d17b8)

# release 0.2.1 (2023-07-31)

* Store pagination to localStorage. (ebf4e9c)
* Fix. Keep page number when switching from one paginator to another. (16ac821)
* Avoid multiple init of a paginator. (87bb51b)
* Add SPaginators class. (ec3673a)
* [refactor] Add filters to Pagination. (4f5daca)
* Move Pagination and initialPagination to ngbrx-paginator.model. (fd02056)
* Remove unused Set Current Paginator action. (ca76eeb)
* Use helper functions in ngbrx-paginator's reducer. (8e60769)

# release 0.2.0 (2023-07-26)

* Update README. (d62e594)
* [BREAKING CHANGE] Modify the structure of NgbrxPaginatorModule.forFeature parameter. (152f9fa)
* Remove unused code. (main) (7f02fc4)

## Breaking changes

The object passed to `NgbrxPaginatorModule.forFeature` has changed. Properties `paginators` and `key` have been removed:

* before:
  ```ts
    NgbrxPaginatorModule.forFeature({
      paginators: [ // you can have as many paginators as you need per module
        {
          // Mandatory
          key: 'Departement/Pagination', // must be unique for the app
          allDataSelector: fromDepartement.selectAll, // ngrx selector returning all the data set

          // Optional
          filters: { // You can provide more than one filter by paginator.
            'Nom': { filter: fromDepartement.byName },
            'Code': { filter: fromDepartement.byCode },
            'Régions/COM': {
              filter: fromDepartement.byRegion,
              values: fromDepartement.selectRegions // Optional: provide select values
            }
          },
          pageSizeOptions: [10, 20, 30] // Defaults to [5, 10, 25, 100]
        }
      ]
    }),
  ```
* after:
  ```ts
    NgbrxPaginatorModule.forFeature({
      'Departement/Pagination': // The name of the paginator must be unique for the application.
      {
        allDataSelector: fromDepartement.selectAll, // @ngrx selector returning all the data set

        // Optional
        filters: { // You can provide more than one filter by paginator.
          'Nom': { filter: fromDepartement.byName },
          'Code': { filter: fromDepartement.byCode },
          'Régions/COM': {
            filter: fromDepartement.byRegion,
            values: fromDepartement.selectRegions // Optional: provide select values
          }
        },
        pageSizeOptions: [10, 20, 30] // Defaults to [5, 10, 25, 100]
      } 
    }),
  ```

# release 0.1.6 (2023-07-25)

* Move toggleSelectedFilterKey click event one element up in the template. (5607a30)

# release 0.1.5 (2023-07-24)

* [temp fix] Force background-color for number of items in paginator filter desc. (HEAD -> main) (00cdbf6)
* Update images for README. (9c1ad1f)

# release 0.1.4 (2023-07-24)

* Add icons to toggle search, lock and unlock filters. (6653056)
* Use button for filter descriptor. (d026f90)
* Remove (click) and (keyup) on default option. (9dd40d7)

# release 0.1.3 (2023-07-18)


projects/test-paginator/src/app/data/departements.ts
import { Departement } from "../departement/departement.model";

export const departements: Departement[] = [
{"code":"01","region":"Auvergne-Rhône-Alpes","chefLieu":"01053","nom":"Ain","typeLiaison":5,"zone":"metro"},
{"code":"02","region":"Hauts-de-France","chefLieu":"02408","nom":"Aisne","typeLiaison":5,"zone":"metro"},
{"code":"03","region":"Auvergne-Rhône-Alpes","chefLieu":"03190","nom":"Allier","typeLiaison":5,"zone":"metro"},
{"code":"04","region":"Provence-Alpes-Côte d'Azur","chefLieu":"04070","nom":"Alpes-de-Haute-Provence","typeLiaison":4,"zone":"metro"},
{"code":"05","region":"Provence-Alpes-Côte d'Azur","chefLieu":"05061","nom":"Hautes-Alpes","typeLiaison":4,"zone":"metro"},
{"code":"06","region":"Provence-Alpes-Côte d'Azur","chefLieu":"06088","nom":"Alpes-Maritimes","typeLiaison":4,"zone":"metro"},
{"code":"07","region":"Auvergne-Rhône-Alpes","chefLieu":"07186","nom":"Ardèche","typeLiaison":5,"zone":"metro"},
{"code":"08","region":"Grand Est","chefLieu":"08105","nom":"Ardennes","typeLiaison":4,"zone":"metro"},
{"code":"09","region":"Occitanie","chefLieu":"09122","nom":"Ariège","typeLiaison":5,"zone":"metro"},
{"code":"10","region":"Grand Est","chefLieu":"10387","nom":"Aube","typeLiaison":5,"zone":"metro"},
{"code":"11","region":"Occitanie","chefLieu":"11069","nom":"Aude","typeLiaison":5,"zone":"metro"},
{"code":"12","region":"Occitanie","chefLieu":"12202","nom":"Aveyron","typeLiaison":5,"zone":"metro"},
{"code":"13","region":"Provence-Alpes-Côte d'Azur","chefLieu":"13055","nom":"Bouches-du-Rhône","typeLiaison":4,"zone":"metro"},
{"code":"14","region":"Normandie","chefLieu":"14118","nom":"Calvados","typeLiaison":2,"zone":"metro"},
{"code":"15","region":"Auvergne-Rhône-Alpes","chefLieu":"15014","nom":"Cantal","typeLiaison":2,"zone":"metro"},
{"code":"16","region":"Nouvelle-Aquitaine","chefLieu":"16015","nom":"Charente","typeLiaison":3,"zone":"metro"},
{"code":"17","region":"Nouvelle-Aquitaine","chefLieu":"17300","nom":"Charente-Maritime","typeLiaison":3,"zone":"metro"},
{"code":"18","region":"Centre-Val de Loire","chefLieu":"18033","nom":"Cher","typeLiaison":2,"zone":"metro"},
{"code":"19","region":"Nouvelle-Aquitaine","chefLieu":"19272","nom":"Corrèze","typeLiaison":3,"zone":"metro"},
{"code":"21","region":"Bourgogne-Franche-Comté","chefLieu":"21231","nom":"Côte-d'Or","typeLiaison":3,"zone":"metro"},
{"code":"22","region":"Bretagne","chefLieu":"22278","nom":"Côtes-d'Armor","typeLiaison":4,"zone":"metro"},
{"code":"23","region":"Nouvelle-Aquitaine","chefLieu":"23096","nom":"Creuse","typeLiaison":3,"zone":"metro"},
{"code":"24","region":"Nouvelle-Aquitaine","chefLieu":"24322","nom":"Dordogne","typeLiaison":3,"zone":"metro"},
{"code":"25","region":"Bourgogne-Franche-Comté","chefLieu":"25056","nom":"Doubs","typeLiaison":2,"zone":"metro"},
{"code":"26","region":"Auvergne-Rhône-Alpes","chefLieu":"26362","nom":"Drôme","typeLiaison":3,"zone":"metro"},
{"code":"27","region":"Normandie","chefLieu":"27229","nom":"Eure","typeLiaison":5,"zone":"metro"},
{"code":"28","region":"Centre-Val de Loire","chefLieu":"28085","nom":"Eure-et-Loir","typeLiaison":1,"zone":"metro"},
{"code":"29","region":"Bretagne","chefLieu":"29232","nom":"Finistère","typeLiaison":2,"zone":"metro"},
{"code":"2A","region":"Corse","chefLieu":"2A004","nom":"Corse-du-Sud","typeLiaison":3,"zone":"metro"},
{"code":"2B","region":"Corse","chefLieu":"2B033","nom":"Haute-Corse","typeLiaison":3,"zone":"metro"},
{"code":"30","region":"Occitanie","chefLieu":"30189","nom":"Gard","typeLiaison":2,"zone":"metro"},
{"code":"31","region":"Occitanie","chefLieu":"31555","nom":"Haute-Garonne","typeLiaison":3,"zone":"metro"},
{"code":"32","region":"Occitanie","chefLieu":"32013","nom":"Gers","typeLiaison":2,"zone":"metro"},
{"code":"33","region":"Nouvelle-Aquitaine","chefLieu":"33063","nom":"Gironde","typeLiaison":3,"zone":"metro"},
{"code":"34","region":"Occitanie","chefLieu":"34172","nom":"Hérault","typeLiaison":5,"zone":"metro"},
{"code":"35","region":"Bretagne","chefLieu":"35238","nom":"Ille-et-Vilaine","typeLiaison":1,"zone":"metro"},
{"code":"36","region":"Centre-Val de Loire","chefLieu":"36044","nom":"Indre","typeLiaison":5,"zone":"metro"},
{"code":"37","region":"Centre-Val de Loire","chefLieu":"37261","nom":"Indre-et-Loire","typeLiaison":1,"zone":"metro"},
{"code":"38","region":"Auvergne-Rhône-Alpes","chefLieu":"38185","nom":"Isère","typeLiaison":5,"zone":"metro"},
{"code":"39","region":"Bourgogne-Franche-Comté","chefLieu":"39300","nom":"Jura","typeLiaison":2,"zone":"metro"},
{"code":"40","region":"Nouvelle-Aquitaine","chefLieu":"40192","nom":"Landes","typeLiaison":4,"zone":"metro"},
{"code":"41","region":"Centre-Val de Loire","chefLieu":"41018","nom":"Loir-et-Cher","typeLiaison":2,"zone":"metro"},
{"code":"42","region":"Auvergne-Rhône-Alpes","chefLieu":"42218","nom":"Loire","typeLiaison":3,"zone":"metro"},
{"code":"43","region":"Auvergne-Rhône-Alpes","chefLieu":"43157","nom":"Haute-Loire","typeLiaison":3,"zone":"metro"},
{"code":"44","region":"Pays de la Loire","chefLieu":"44109","nom":"Loire-Atlantique","typeLiaison":3,"zone":"metro"},
{"code":"45","region":"Centre-Val de Loire","chefLieu":"45234","nom":"Loiret","typeLiaison":2,"zone":"metro"},
{"code":"46","region":"Occitanie","chefLieu":"46042","nom":"Lot","typeLiaison":2,"zone":"metro"},
{"code":"47","region":"Nouvelle-Aquitaine","chefLieu":"47001","nom":"Lot-et-Garonne","typeLiaison":2,"zone":"metro"},
{"code":"48","region":"Occitanie","chefLieu":"48095","nom":"Lozère","typeLiaison":3,"zone":"metro"},
{"code":"49","region":"Pays de la Loire","chefLieu":"49007","nom":"Maine-et-Loire","typeLiaison":0,"zone":"metro"},
{"code":"50","region":"Normandie","chefLieu":"50502","nom":"Manche","typeLiaison":3,"zone":"metro"},
{"code":"51","region":"Grand Est","chefLieu":"51108","nom":"Marne","typeLiaison":3,"zone":"metro"},
{"code":"52","region":"Grand Est","chefLieu":"52121","nom":"Haute-Marne","typeLiaison":3,"zone":"metro"},
{"code":"53","region":"Pays de la Loire","chefLieu":"53130","nom":"Mayenne","typeLiaison":3,"zone":"metro"},
{"code":"54","region":"Grand Est","chefLieu":"54395","nom":"Meurthe-et-Moselle","typeLiaison":0,"zone":"metro"},
{"code":"55","region":"Grand Est","chefLieu":"55029","nom":"Meuse","typeLiaison":3,"zone":"metro"},
{"code":"56","region":"Bretagne","chefLieu":"56260","nom":"Morbihan","typeLiaison":2,"zone":"metro"},
{"code":"57","region":"Grand Est","chefLieu":"57463","nom":"Moselle","typeLiaison":3,"zone":"metro"},
{"code":"58","region":"Bourgogne-Franche-Comté","chefLieu":"58194","nom":"Nièvre","typeLiaison":3,"zone":"metro"},
{"code":"59","region":"Hauts-de-France","chefLieu":"59350","nom":"Nord","typeLiaison":2,"zone":"metro"},
{"code":"60","region":"Hauts-de-France","chefLieu":"60057","nom":"Oise","typeLiaison":5,"zone":"metro"},
{"code":"61","region":"Normandie","chefLieu":"61001","nom":"Orne","typeLiaison":5,"zone":"metro"},
{"code":"62","region":"Hauts-de-France","chefLieu":"62041","nom":"Pas-de-Calais","typeLiaison":2,"zone":"metro"},
{"code":"63","region":"Auvergne-Rhône-Alpes","chefLieu":"63113","nom":"Puy-de-Dôme","typeLiaison":2,"zone":"metro"},
{"code":"64","region":"Nouvelle-Aquitaine","chefLieu":"64445","nom":"Pyrénées-Atlantiques","typeLiaison":4,"zone":"metro"},
{"code":"65","region":"Occitanie","chefLieu":"65440","nom":"Hautes-Pyrénées","typeLiaison":4,"zone":"metro"},
{"code":"66","region":"Occitanie","chefLieu":"66136","nom":"Pyrénées-Orientales","typeLiaison":4,"zone":"metro"},
{"code":"67","region":"Grand Est","chefLieu":"67482","nom":"Bas-Rhin","typeLiaison":2,"zone":"metro"},
{"code":"68","region":"Grand Est","chefLieu":"68066","nom":"Haut-Rhin","typeLiaison":2,"zone":"metro"},
{"code":"69","region":"Auvergne-Rhône-Alpes","chefLieu":"69123","nom":"Rhône","typeLiaison":2,"zone":"metro"},
{"code":"70","region":"Bourgogne-Franche-Comté","chefLieu":"70550","nom":"Haute-Saône","typeLiaison":3,"zone":"metro"},
{"code":"71","region":"Bourgogne-Franche-Comté","chefLieu":"71270","nom":"Saône-et-Loire","typeLiaison":0,"zone":"metro"},
{"code":"72","region":"Pays de la Loire","chefLieu":"72181","nom":"Sarthe","typeLiaison":3,"zone":"metro"},
{"code":"73","region":"Auvergne-Rhône-Alpes","chefLieu":"73065","nom":"Savoie","typeLiaison":3,"zone":"metro"},
{"code":"74","region":"Auvergne-Rhône-Alpes","chefLieu":"74010","nom":"Haute-Savoie","typeLiaison":3,"zone":"metro"},
{"code":"75","region":"Ile-de-France","chefLieu":"75056","nom":"Paris","typeLiaison":0,"zone":"metro"},
{"code":"76","region":"Normandie","chefLieu":"76540","nom":"Seine-Maritime","typeLiaison":3,"zone":"metro"},
{"code":"77","region":"Ile-de-France","chefLieu":"77288","nom":"Seine-et-Marne","typeLiaison":0,"zone":"metro"},
{"code":"78","region":"Ile-de-France","chefLieu":"78646","nom":"Yvelines","typeLiaison":4,"zone":"metro"},
{"code":"79","region":"Nouvelle-Aquitaine","chefLieu":"79191","nom":"Deux-Sèvres","typeLiaison":4,"zone":"metro"},
{"code":"80","region":"Hauts-de-France","chefLieu":"80021","nom":"Somme","typeLiaison":3,"zone":"metro"},
{"code":"81","region":"Occitanie","chefLieu":"81004","nom":"Tarn","typeLiaison":2,"zone":"metro"},
{"code":"82","region":"Occitanie","chefLieu":"82121","nom":"Tarn-et-Garonne","typeLiaison":2,"zone":"metro"},
{"code":"83","region":"Provence-Alpes-Côte d'Azur","chefLieu":"83137","nom":"Var","typeLiaison":2,"zone":"metro"},
{"code":"84","region":"Provence-Alpes-Côte d'Azur","chefLieu":"84007","nom":"Vaucluse","typeLiaison":2,"zone":"metro"},
{"code":"85","region":"Pays de la Loire","chefLieu":"85191","nom":"Vendée","typeLiaison":3,"zone":"metro"},
{"code":"86","region":"Nouvelle-Aquitaine","chefLieu":"86194","nom":"Vienne","typeLiaison":3,"zone":"metro"},
{"code":"87","region":"Nouvelle-Aquitaine","chefLieu":"87085","nom":"Haute-Vienne","typeLiaison":3,"zone":"metro"},
{"code":"88","region":"Grand Est","chefLieu":"88160","nom":"Vosges","typeLiaison":4,"zone":"metro"},
{"code":"89","region":"Bourgogne-Franche-Comté","chefLieu":"89024","nom":"Yonne","typeLiaison":5,"zone":"metro"},
{"code":"90","region":"Bourgogne-Franche-Comté","chefLieu":"90010","nom":"Territoire de Belfort","typeLiaison":2,"zone":"metro"},
{"code":"91","region":"Ile-de-France","chefLieu":"91228","nom":"Essonne","typeLiaison":5,"zone":"metro"},
{"code":"92","region":"Ile-de-France","chefLieu":"92050","nom":"Hauts-de-Seine","typeLiaison":4,"zone":"metro"},
{"code":"93","region":"Ile-de-France","chefLieu":"93008","nom":"Seine-Saint-Denis","typeLiaison":3,"zone":"metro"},
{"code":"94","region":"Ile-de-France","chefLieu":"94028","nom":"Val-de-Marne","typeLiaison":2,"zone":"metro"},
{"code":"95","region":"Ile-de-France","chefLieu":"95500","nom":"Val-d'Oise","typeLiaison":2,"zone":"metro"},
{"code":"971","region":"Guadeloupe","chefLieu":"97105","nom":"Guadeloupe","typeLiaison":3,"zone":"drom"},
{"code":"972","region":"Martinique","chefLieu":"97209","nom":"Martinique","typeLiaison":3,"zone":"drom"},
{"code":"973","region":"Guyane","chefLieu":"97302","nom":"Guyane","typeLiaison":3,"zone":"drom"},
{"code":"974","region":"La Réunion","chefLieu":"97411","nom":"La Réunion","typeLiaison":0,"zone":"drom"},
{"code":"976","region":"Mayotte","chefLieu":"97608","nom":"Mayotte","typeLiaison":0,"zone":"drom"},
{"code":"975","region":"Collectivité d'Outre-mer","chefLieu":"97502","nom":"Saint-Pierre-et-Miquelon","typeLiaison":0,"zone":"com"},
{"code":"977","region":"Collectivité d'Outre-mer","chefLieu":"97701","nom":"Saint-Barthélemy","typeLiaison":0,"zone":"com"},
{"code":"978","region":"Collectivité d'Outre-mer","chefLieu":"97801","nom":"Saint-Martin","typeLiaison":0,"zone":"com"},
{"code":"984","region":"Collectivité d'Outre-mer","chefLieu":"97502","nom":"Terres australes et antarctiques françaises","typeLiaison":0,"zone":"com"},
{"code":"986","region":"Collectivité d'Outre-mer","chefLieu":"98613","nom":"Wallis et Futuna","typeLiaison":0,"zone":"com"},
{"code":"987","region":"Collectivité d'Outre-mer","chefLieu":"98735","nom":"Polynésie française","typeLiaison":0,"zone":"com"},
{"code":"988","region":"Collectivité d'Outre-mer","chefLieu":"98818","nom":"Nouvelle-Calédonie","typeLiaison":0,"zone":"com"},
{"code":"989","region":"Collectivité d'Outre-mer","chefLieu":"98901","nom":"Île de Clipperton","typeLiaison":0,"zone":"com"}
]
